src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+
+      val jobUniqueId = Math.abs(UUID.randomUUID.getLeastSignificantBits.toInt)
+      val driverHostAddress = ClusterUtil.getDriverHost(df)
+      val port = spanningTree.getPort
+
+      /*
+      --span_server specifies the network address of a little server that sets up spanning trees over the nodes.
+      --unique_id should be a number that is the same for all nodes executing a particular job and
+        different for all others.
+      --total is the total number of nodes.
+      --node should be unique for each node and range from {0,total-1}.
+      --holdout_off should be included for distributed training
+      */
+      vwArgs.append(s" --holdout_off --span_server $driverHostAddress --span_server_port $port ")
+        .append(s"--unique_id $jobUniqueId --total $numWorkers ")
+
+      trainInternal(df, vwArgs.result, () => s"--node ${TaskContext.get.partitionId}")
+    } finally {
+      spanningTree.stop
+    }
+  }
+
+  /**
+    * Main training loop
+    * @param dataset input data.
+    * @return binary VW model.
+    */
+  protected def trainInternal(dataset: Dataset[_]): Array[Byte] = {
+    // follow LightGBM pattern
+    val numCoresPerExec = ClusterUtil.getNumCoresPerExecutor(dataset)
+    val numExecutorCores = ClusterUtil.getNumExecutorCores(dataset, numCoresPerExec)
+    val numWorkers = min(numExecutorCores, dataset.rdd.getNumPartitions)
+
+    // Reduce number of partitions to number of executor cores
+    val df = if (dataset.rdd.getNumPartitions > numWorkers)
+        dataset.toDF.coalesce(numWorkers)
+      else
+        dataset.toDF
+
+    // add exposed parameters to the final command line args
+    val vwArgs = new StringBuilder()
+      .append(s"${getArgs} --save_resume --preserve_performance_counters ")
+      .append(s"--hash_seed ${getHashSeed} ")
+      .append(s"-b ${getNumBits} ")
+      .appendParam(learningRate, "-l")
+      .appendParam(powerT, "--power_t")
+      .appendParam(l1, "--l1")
+      .appendParam(l2, "--l2")
+      .appendParam(ignoreNamespaces, "--ignore")
+      .appendParam(interactions, "-q")
+
+    // bfgs needs cache file
+    if (getArgs.contains("--bfgs"))
+      setEnableCacheFile(true)
+
+    // call training
+    val binaryModel = if (numWorkers == 1)
+      trainInternal(df, vwArgs.result)
+    else
+      trainInternalDistributed(df, vwArgs, numWorkers)
+
+    binaryModel.get
+  }
+}
+
+/**
+  * Base trait to wrap the model for prediction.
+  */
+trait VowpalWabbitBaseModel extends org.apache.spark.ml.param.shared.HasFeaturesCol
+  with org.apache.spark.ml.param.shared.HasRawPredictionCol
+{
+  /**
+    * The serialized VW model.
+    */
+  val model: Array[Byte]
+
+  @transient
+  lazy val vw = new VowpalWabbitNative("--testonly --quiet", model)
+
+  @transient
+  lazy val example = vw.createExample()
+
+  @transient
+  lazy val vwArgs = vw.getArguments
+
+  val additionalFeatures = new StringArrayParam(this, "additionalFeatures", "Additional feature columns")
```

Comment:
how does this param work?  If this is letting some additional cols pass through, I don't think it fits in with spark ML.  with predict, all original columns should be passed through with the additional ones added.

Reply:
when one wants to create quadratic features within VW you'd specify additionalFeatures. each feature column is treated as one namespace. Using -q 'uc' for columns 'user' and 'content' you'd get all quadratics for features in user/content (the first letter is called feature group and VW users are used to it... before somebody starts complaining ;))